src/app/Events/EventTypes.tsx

```diff
@@ -0,0 +1,140 @@
+import * as React from 'react';
+import { filter, map } from 'rxjs/operators';
+import { Pagination, Split, SplitItem, TextInput } from '@patternfly/react-core';
+import { Table, TableBody, TableHeader, TableVariant, expandable } from '@patternfly/react-table';
+import { ServiceContext } from '@app/Shared/Services/Services';
+
+export interface EventType {
+  name: string;
+  typeId: string;
+  description: string;
+  category: string[];
+  options: { [key: string]: OptionDescriptor }[];
+}
+
+export interface OptionDescriptor {
+  name: string;
+  description: string;
+  defaultValue: string;
+}
+
+export const EventTypes = (props) => {
+  const context = React.useContext(ServiceContext);
+
+  const [types, setTypes] = React.useState([] as EventType[]);
+  const [displayedTypes, setDisplayedTypes] = React.useState([] as any[]);
+  const [currentPage, setCurrentPage] = React.useState(1);
+  const [perPage, setPerPage] = React.useState(10);
+  const [openRow, setOpenRow] = React.useState(-1);
+  const [filterText, setFilterText] = React.useState('');
+
+  const tableColumns = [
+    {
+      title: 'Name',
+      cellFormatters: [expandable]
+    },
+    'Type ID',
+    'Description',
+    'Categories'
+  ];
+
+  React.useEffect(() => {
+    const sub = context.commandChannel.onResponse('list-event-types')
+      .pipe(
+        filter(m => m.status === 0),
+        map(m => m.payload),
+      )
+      .subscribe(types => setTypes(types));
+    return () => sub.unsubscribe();
+  }, []);
+
+  React.useEffect(() => {
+    context.commandChannel.sendMessage('list-event-types');
+  }, []);
+
+  React.useEffect(() => {
+    const offset = (currentPage - 1) * perPage;
+    const page = filterTypesByText(types, filterText).slice(offset, offset + perPage);
+
+    const rows: any[] = [];
+    page.forEach((t: EventType, idx: number) => {
+      rows.push({ cells: [ t.name, t.typeId, t.description, t.category.join(', ').trim() ], isOpen: (idx === openRow) });
+      if (idx === openRow) {
+        let child = '';
+        for (const opt in t.options) {
+          child += `${opt}=[${t.options[opt].defaultValue}]\t`;
+        }
+        rows.push({ parent: idx, fullWidth: true, cells: [ child ] });
+      }
+    });
+
+    setDisplayedTypes(rows);
+  }, [currentPage, perPage, types, openRow, filterText]);
+
+  const filterTypesByText = (types, filter) => {
+    if (!filter) {
+      return types;
+    }
+    const includesSubstr = (a, b) => !!a && !!b && a.toLowerCase().includes(b.trim().toLowerCase());
+    return types.filter(t => {
+      if (includesSubstr(t.name, filter)) {
+        return true;
+      }
+      if (includesSubstr(t.typeId, filter)) {
+        return true;
+      }
+      if (includesSubstr(t.description, filter)) {
+        return true;
+      }
+      if (t.category.some(c => includesSubstr(c, filter))) {
```

Comment:
This filter setup doesn't seem to work for targeting multiple categories, for example, both 'Java Virtual Machine' and 'GC'. My attempts were "Java Virtual Machine, GC" && "Java Virtual Machine GC". The categories are displayed as a comma separated list, so I'd hope the comma separated list filter would work too.

Reply:
Right, it doesn't. It's just a simple substring match, so you get an implicit `or` of categories anyway, ie filtering will narrow down the list as long as the event has at least one matching category.

Reply:
Are you trying to get it to do an `and` filtering? That would probably require a more complex filtering setup than the current simple text box. It could be filed as a future enhancement.

Reply:
Hmm, well the category list displays text e.g. 'Java Development Kit, Security'

It's unclear to a user why the same text in the filter would not match anything. I think the most simple pass would also check against the text displayed, such that the comma-separated list would work. Maybe you can construct that for checking against or save it somewhere to be reused?

Reply:
Basically, without any use of explicit && || type operators in the text box, I'd expect it to be a text search, with the text that is displayed to me being searched.

Reply:
Sure, I could do that, but it wouldn't match "Security, Java Development Kit" if some other event happens to have that list of categories (same members, different order).

Reply:
(although searching for each substring separately would of course still work)

Reply:
I think that sounds fine for now. The category text looks sorted fortunately, so at least there won't be event type with Security & JDK, but in different order.

Reply:
That would match how basic text search works anyways I think?